packages/react-jss/src/create-use-styles.js

```diff
@@ -0,0 +1,101 @@
+// @flow
+
+import React from 'react'
+import {ThemeContext as DefaultThemeContext} from 'theming'
+
+import {JssContext} from './JssContext'
+import {
+  createStaticSheet,
+  addDynamicRules,
+  removeDynamicRules,
+  updateDynamicRules
+} from './utils/sheets'
+import {getIndex} from './utils/index-counter'
+import type {HookOptions, Styles} from './types'
+import {unmanageSheet, manageSheet} from './utils/managers'
+import {getSheetClasses} from './utils/get-sheet-classes'
+
+const noTheme = {}
+
+function createUseStyles<Theme: {}>(styles: Styles<Theme>, options?: HookOptions<Theme> = {}) {
+  const {index = getIndex(), theming, name = 'Hook', ...sheetOptions} = options
+  const ThemeContext = (theming && theming.context) || DefaultThemeContext
+
+  return (data: any) => {
+    const jssContext = React.useContext(JssContext)
+    // $FlowFixMe
+    const theme: Theme = typeof styles === 'function' ? React.useContext(ThemeContext) : noTheme
+
+    // When the theme or the context changes we create a new sheet
+    const sheet = React.useMemo(
+      () => {
+        if (jssContext.disableStylesGeneration) {
+          return undefined
+        }
+
+        const stylesheet = createStaticSheet({
+          context: jssContext,
+          styles,
+          name,
+          theme,
+          index,
+          sheetOptions
+        })
+
+        manageSheet({
+          index,
+          context: jssContext,
+          sheet: stylesheet,
+          theme
+        })
+
+        return sheet
+      },
+      [theme, jssContext]
+    )
+
+    // When the sheet changes, we readd the dynamic rules and update them
+    const dynamicRules = React.useMemo(
+      () => {
+        const rules = addDynamicRules(sheet)
+
+        updateDynamicRules(data, sheet, rules)
+
+        return rules
+      },
+      [sheet]
+    )
+
+    // Update the dynamic rules before the actual render if the data has changed
+    React.useLayoutEffect(
```

Comment:
It needs to be changed for SSR.

Reply:
I'm not sure if we can get around that. I think we need `useLayoutEffect` to flush the changes before the actual render happens, so we don't cause an additional render.

I'm not sure why react warns in this case as this is a valid use case.

Reply:
React doesn't know that the server rendered output will include the styling so they assume this causes flashes. 

I nasty workaround is to use something like `const useEnhancedEffect = onServer ? useEffect : useLayoutEffect` which gets rid of the warning but looses the linter.

Reply:
This workaround seems to fit accurately to what's going on, linter can be disabled for that line specifically.